File: 994.Rotting-Oranges.go
// https://leetcode.com/problems/rotting-oranges/
//
// algorithms
// Easy (46.52%)
// Total Accepted:    3,200
// Total Submissions: 6,879
// beats 100.0% of golang submissions

package leetcode

func orangesRotting(grid [][]int) int {
	row, col := len(grid), len(grid[0])
	var rottenOranges [][]int

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if grid[i][j] == 2 {
				rottenOranges = append(rottenOranges, []int{i, j})
			}
		}
	}

	var rotNum, times int

	for {
		rotNum = 0
		length := len(rottenOranges)
		var tmp [][]int
		for i := 0; i < length; i++ {
			x, y := rottenOranges[i][0], rottenOranges[i][1]
			for _, p := range [][]int{{x - 1, y}, {x + 1, y}, {x, y - 1}, {x, y + 1}} {
				if p[0] < 0 || p[0] >= row || p[1] < 0 || p[1] >= col {
					continue
				}
				if grid[p[0]][p[1]] == 1 {
					grid[p[0]][p[1]] = 2
					rotNum++
					tmp = append(tmp, []int{p[0], p[1]})
				}
			}
		}
		if rotNum == 0 {
			break
		}
		times++
		rottenOranges = tmp
	}

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if grid[i][j] == 1 {
				return -1
			}
		}
	}

	return times
}

File: 994.Rotting-Oranges.java
// https://leetcode.com/problems/rotting-oranges/
//
// algorithms
// Easy (46.61%)
// Total Accepted: 32,491
// Total Submissions: 69,714
// beats 100.0% of java submissions


class Solution {

    private static final int FRESH = 1;
    private static final int ROT = 2;

    public int orangesRotting(int[][] grid) {
        int row = grid.length, col = grid[0].length;

        int initFreshNum = 0;
        LinkedList<int[]> q = new LinkedList<>();
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == ROT) {
                    q.offer(new int[] { i, j });
                } else if (grid[i][j] == FRESH) {
                    initFreshNum++;
                }
            }
        }
        if (initFreshNum == 0) {
            return 0;
        }

        int minute = 0, rotNum = 0;
        while (!q.isEmpty()) {
            LinkedList<int[]> tmp = new LinkedList<>();
            boolean freshFound = false;

            while (!q.isEmpty()) {
                int[] cur = q.pollFirst();

                if (cur[0] > 0 && grid[cur[0] - 1][cur[1]] == FRESH) {
                    rotNum++;
                    tmp.offer(new int[] { cur[0] - 1, cur[1] });
                    grid[cur[0] - 1][cur[1]] = ROT;
                    freshFound = true;
                }
                if (cur[0] < row - 1 && grid[cur[0] + 1][cur[1]] == FRESH) {
                    rotNum++;
                    tmp.offer(new int[] { cur[0] + 1, cur[1] });
                    grid[cur[0] + 1][cur[1]] = ROT;
                    freshFound = true;
                }
                if (cur[1] > 0 && grid[cur[0]][cur[1] - 1] == FRESH) {
                    rotNum++;
                    tmp.offer(new int[] { cur[0], cur[1] - 1 });
                    grid[cur[0]][cur[1] - 1] = ROT;
                    freshFound = true;
                }
                if (cur[1] < col - 1 && grid[cur[0]][cur[1] + 1] == FRESH) {
                    rotNum++;
                    tmp.offer(new int[] { cur[0], cur[1] + 1 });
                    grid[cur[0]][cur[1] + 1] = ROT;
                    freshFound = true;
                }
            }

            q = tmp;
            if (freshFound) {
                minute++;
            }
        }

        if (rotNum < initFreshNum) {
            return -1;
        }

        return minute;
    }
}